airbyte-integrations/connectors/source-linnworks/source_linnworks/streams.py

```diff
@@ -0,0 +1,227 @@
+    # https://apps.linnworks.net/Api/Class/linnworks-spa-commondata-Generic-GenericPagedResult
+    @abstractmethod
+    def paged_result(self, response: requests.Response) -> Mapping[str, Any]:
+        pass
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        result = self.paged_result(response)
+
+        if result["PageNumber"] < result["TotalPages"]:
+            return {
+                "PageNumber": result["PageNumber"] + 1,
+                "EntriesPerPage": result["EntriesPerPage"],
+                "TotalEntries": result["TotalEntries"],
+                "TotalPages": result["TotalPages"],
+            }
+
+
+class Location(LinnworksStream):
+    # https://apps.linnworks.net/Api/Method/Locations-GetLocation
+    # Response: StockLocation https://apps.linnworks.net/Api/Class/linnworks-spa-commondata-Locations-ClassBase-StockLocation
+    # Allows 150 calls per minute
+    primary_key = "StockLocationIntId"
+
+    def path(
+        self, stream_state: Mapping[str, Any] = None, stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> str:
+        return "/api/Locations/GetLocation"
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        return stream_state
+
+
+class StockLocations(LinnworksStream):
+    # https://apps.linnworks.net/Api/Method/Inventory-GetStockLocations
+    # Response: List<StockLocation> https://apps.linnworks.net/Api/Class/linnworks-spa-commondata-Inventory-ClassBase-StockLocation
+    # Allows 150 calls per minute
+    primary_key = "StockLocationIntId"
+
+    def path(
+        self, stream_state: Mapping[str, Any] = None, stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> str:
+        return "/api/Inventory/GetStockLocations"
+
+    def read_records(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_slice: Mapping[str, Any] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping[str, Any]]:
+        records = super().read_records(sync_mode, cursor_field, stream_slice, stream_state)
+
+        for record in records:
+            location = Location(authenticator=self.authenticator)
+            srecords = location.read_records(sync_mode, cursor_field, stream_slice, {"pkStockLocationId": record["StockLocationId"]})
+            record["location"] = next(srecords)
+            yield record
+
+
+class StockItems(LinnworksStream):
+    # https://apps.linnworks.net//Api/Method/Stock-GetStockItemsFull
+    # Response: List<StockItemFull> https://apps.linnworks.net/Api/Class/linnworks-spa-commondata-Inventory-ClassBase-StockItemFull
+    # Allows 250 calls per minute
+    primary_key = "StockItemIntId"
+    page_size = 200
+
+    raise_on_http_errors = False
```

Comment:
As I mentioned in `bootstrap.md`, `GetStockItemsFull` is an endpoint that uses ad-hoc pagination instead of `GenericPagedResult`. The former doesn't report ahead of time the number of total items or total pages. 

Let's say we have `201` items and set `page_size=100`, then the paged request will look like this:

- `/api/Stock/GetStockItemsFull?entriesPerPage=100&pageNumber=1`
- `/api/Stock/GetStockItemsFull?entriesPerPage=100&pageNumber=2`
- `/api/Stock/GetStockItemsFull?entriesPerPage=100&pageNumber=3`

The last request will return 1 item, and since we know the page size, we're safe to stop here.

The issue appears when the number of items is multiple of `page_size`. The same example with `200` items and `page_size=100`:

- `/api/Stock/GetStockItemsFull?entriesPerPage=100&pageNumber=1`
- `/api/Stock/GetStockItemsFull?entriesPerPage=100&pageNumber=2`
- `/api/Stock/GetStockItemsFull?entriesPerPage=100&pageNumber=3` **// ERROR, no more items**

If `raise_on_http_errors` is set to `True`, this false positive HTTP error propagates to the stream's output.